core/amber/src/main/scala/edu/uci/ics/texera/web/resource/aiassistant/python_abstract_syntax_tree.py

```diff
@@ -0,0 +1,39 @@
+import ast
+import json
+import sys
+import base64
+
+class TypeAnnotationVisitor(ast.NodeVisitor):
+    def __init__(self, start_line_offset=0):
+        self.untyped_args = []
+        # To calculate the correct start line
+        self.start_line_offset = start_line_offset
+
+    def visit_FunctionDef(self, node):
+        for arg in node.args.args:
+            # Self is not an argument
+            if arg.arg == 'self':
+                continue
+            if not arg.annotation:
+                # +1 and -1 is to change ast line/col to monaco editor line/col for the range
+                start_line = arg.lineno + self.start_line_offset - 1
+                start_col = arg.col_offset + 1
+                end_line = start_line
+                end_col = start_col + len(arg.arg)
```

Comment:
I am very concerned that you are using starting line, col to do such change in the text level. Think of this as a compiler point of view, the input code could be very complex (e.g., arguments spanning multiple lines). What you should do instead, is to introdue a new AST node (or modify an existing AST node) to add the annotation, and let AST build the new code for you. 

AFAIK, the arguments in AST have the following structure (copied from AST doc):
```
class ast.arg(arg, annotation, type_comment)
A single argument in a list. arg is a raw string of the argument name; annotation is its annotation, such as a Name node.
```

Reply:
I have tested the code with arguments spanning multiple lines, and the functionalities work correctly. The startline, startcolumn, endline, and endcolumn accurately locate the code within the UDF. The main reason I chose to use startline, etc., is that the range in Monaco Editor differs from the range in the AST. If you try to use the AST to find an argument and instruct Monaco Editor to insert the argument right after the AST range, it will end up inserting it in the wrong place. In my implementation, the startline and startcolumn undergo minor adjustments before being wrapped as a Monaco Editor range to ensure the annotation is inserted in the correct location. 
```
                start_line = arg.lineno + self.start_line_offset - 1
                start_col = arg.col_offset + 1
```
Additionally, using the range directly returned from the AST can cause highlighting errors for arguments.

Reply:
Maybe just include a test case to show what you said? If an offline discussion is easier, we can talk.

Reply:
I will record a gif later to show you.

Reply:
Another thing is, maybe you are stuck with thinking of "inserting the annotations" in the Monaco editor. An alternative way, is to let ADT generate the entire new code snippet with annotations inserted, then on Monaco simply replace the old code snippet with the new generated one. This way you don't have to deal with inserting into Monaco.

Reply:
Yes, that is a way I originally tried but gave up. A key reason we can't get rid of the startline, endline, etc. is because Monaco Editor needs the range to highlight the argument that we are giving annotation suggestion, and the range is composed of startline, endline, and stuff.